File: backend/internal/models/lab.go
package models

import (
	"encoding/json"
	"time"
)

// LabStrategyEntryName is the canonical display name for the lab-generated entry
// when it participates in simulated calcutta evaluations.
const LabStrategyEntryName = "Our Strategy"

// InvestmentModel represents a lab.investment_models row.
type InvestmentModel struct {
	ID           string          `json:"id"`
	Name         string          `json:"name"`
	Kind         string          `json:"kind"`
	ParamsJSON   json.RawMessage `json:"paramsJson"`
	Notes        *string         `json:"notes,omitempty"`
	CreatedAt    time.Time       `json:"createdAt"`
	UpdatedAt    time.Time       `json:"updatedAt"`
	NEntries     int             `json:"nEntries"`
	NEvaluations int             `json:"nEvaluations"`
}

// LabEntry represents a lab.entries row.
type LabEntry struct {
	ID                    string          `json:"id"`
	InvestmentModelID     string          `json:"investmentModelId"`
	CalcuttaID            string          `json:"calcuttaId"`
	GameOutcomeKind       string          `json:"gameOutcomeKind"`
	GameOutcomeParamsJSON json.RawMessage `json:"gameOutcomeParamsJson"`
	OptimizerKind         string          `json:"optimizerKind"`
	OptimizerParamsJSON   json.RawMessage `json:"optimizerParamsJson"`
	StartingStateKey      string          `json:"startingStateKey"`
	PredictionsJSON       json.RawMessage `json:"predictionsJson,omitempty"`
	BidsJSON              json.RawMessage `json:"bidsJson"`
	CreatedAt             time.Time       `json:"createdAt"`
	UpdatedAt             time.Time       `json:"updatedAt"`
}

// LabEntryDetail is LabEntry with joined data for display.
type LabEntryDetail struct {
	LabEntry
	ModelName    string `json:"modelName"`
	ModelKind    string `json:"modelKind"`
	CalcuttaName string `json:"calcuttaName"`
	NEvaluations int    `json:"nEvaluations"`
}

// LabPrediction represents a single market prediction within predictions_json.
type LabPrediction struct {
	TeamID               string  `json:"teamId"`
	PredictedMarketShare float64 `json:"predictedMarketShare"`
	ExpectedPoints       float64 `json:"expectedPoints"`
}

// LabEnrichedPrediction is LabPrediction with team details for display.
type LabEnrichedPrediction struct {
	TeamID               string  `json:"teamId"`
	SchoolName           string  `json:"schoolName"`
	Seed                 int     `json:"seed"`
	Region               string  `json:"region"`
	PredictedMarketShare float64 `json:"predictedMarketShare"`
	PredictedBidPoints   int     `json:"predictedBidPoints"`
	ExpectedPoints       float64 `json:"expectedPoints"`
	ExpectedROI          float64 `json:"expectedRoi"`
	RationalPoints       int     `json:"rationalPoints"`
	EdgePercent          float64 `json:"edgePercent"`
}

// LabEntryBid represents a single optimized bid within bids_json.
type LabEntryBid struct {
	TeamID      string   `json:"teamId"`
	BidPoints   int      `json:"bidPoints"`
	ExpectedROI *float64 `json:"expectedRoi,omitempty"`
}

// LabEnrichedBid is LabEntryBid with team details and rational allocation for display.
type LabEnrichedBid struct {
	TeamID         string   `json:"teamId"`
	SchoolName     string   `json:"schoolName"`
	Seed           int      `json:"seed"`
	Region         string   `json:"region"`
	BidPoints      int      `json:"bidPoints"`
	RationalPoints int      `json:"rationalPoints"`
	EdgePercent    float64  `json:"edgePercent"`
	ExpectedROI    *float64 `json:"expectedRoi,omitempty"`
}

// LabTeamInfo holds team metadata used during enrichment.
type LabTeamInfo struct {
	Name   string
	Seed   int
	Region string
}

// LabEntryRaw holds the raw entry data fetched from the database before enrichment.
type LabEntryRaw struct {
	ID                    string
	InvestmentModelID     string
	CalcuttaID            string
	GameOutcomeKind       string
	GameOutcomeParamsJSON json.RawMessage
	OptimizerKind         string
	OptimizerParamsJSON   json.RawMessage
	StartingStateKey      string
	CreatedAt             time.Time
	UpdatedAt             time.Time
	ModelName             string
	ModelKind             string
	CalcuttaName          string
	NEvaluations          int
	HasPredictions        bool
	Predictions           []LabPrediction
	Bids                  []LabEntryBid
	Teams                 map[string]LabTeamInfo
	TotalPoolBudget       int
}

// LabEntryDetailEnriched is LabEntryDetail with enriched predictions and bids.
type LabEntryDetailEnriched struct {
	ID                    string                 `json:"id"`
	InvestmentModelID     string                 `json:"investmentModelId"`
	CalcuttaID            string                 `json:"calcuttaId"`
	GameOutcomeKind       string                 `json:"gameOutcomeKind"`
	GameOutcomeParamsJSON json.RawMessage        `json:"gameOutcomeParamsJson"`
	OptimizerKind         string                 `json:"optimizerKind"`
	OptimizerParamsJSON   json.RawMessage        `json:"optimizerParamsJson"`
	StartingStateKey      string                 `json:"startingStateKey"`
	CreatedAt             time.Time              `json:"createdAt"`
	UpdatedAt             time.Time              `json:"updatedAt"`
	ModelName             string                 `json:"modelName"`
	ModelKind             string                 `json:"modelKind"`
	CalcuttaName          string                 `json:"calcuttaName"`
	NEvaluations          int                    `json:"nEvaluations"`
	HasPredictions        bool                   `json:"hasPredictions"`
	Predictions           []LabEnrichedPrediction `json:"predictions,omitempty"`
	Bids                  []LabEnrichedBid        `json:"bids"`
}

// LabEvaluation represents a lab.evaluations row.
type LabEvaluation struct {
	ID                     string          `json:"id"`
	EntryID                string          `json:"entryId"`
	NSims                  int             `json:"nSims"`
	Seed                   int             `json:"seed"`
	MeanNormalizedPayout   *float64        `json:"meanNormalizedPayout,omitempty"`
	MedianNormalizedPayout *float64        `json:"medianNormalizedPayout,omitempty"`
	PTop1                  *float64        `json:"pTop1,omitempty"`
	PInMoney               *float64        `json:"pInMoney,omitempty"`
	OurRank                *int            `json:"ourRank,omitempty"`
	SummaryJSON            json.RawMessage `json:"summaryJson,omitempty"`
	CreatedAt              time.Time       `json:"createdAt"`
	UpdatedAt              time.Time       `json:"updatedAt"`
}

// LabEvaluationSummary is the structured narrative computed for an evaluation.
type LabEvaluationSummary struct {
	PercentileRank float64                       `json:"percentileRank"`
	VsBaseline     *LabEvaluationBaselineComparison `json:"vsBaseline,omitempty"`
	NEntries       int                           `json:"nEntries"`
	TopHoldings    []LabEvaluationTopHolding     `json:"topHoldings"`
	KeyInsight     string                        `json:"keyInsight"`
}

// LabEvaluationBaselineComparison compares an evaluation to a naive_ev baseline.
type LabEvaluationBaselineComparison struct {
	MeanPayoutDelta float64 `json:"meanPayoutDelta"`
	PTop1Delta      float64 `json:"pTop1Delta"`
	Interpretation  string  `json:"interpretation"`
}

// LabEvaluationTopHolding represents a top bid by points in the strategy.
type LabEvaluationTopHolding struct {
	SchoolName string `json:"schoolName"`
	Seed       int    `json:"seed"`
	BidPoints  int    `json:"bidPoints"`
}

// LabEvaluationDetail is LabEvaluation with joined data for display.
type LabEvaluationDetail struct {
	LabEvaluation
	ModelName        string `json:"modelName"`
	ModelKind        string `json:"modelKind"`
	CalcuttaID       string `json:"calcuttaId"`
	CalcuttaName     string `json:"calcuttaName"`
	StartingStateKey string `json:"startingStateKey"`
}

// LabEvaluationEntryResult represents a single entry's result in an evaluation.
type LabEvaluationEntryResult struct {
	ID                   string   `json:"id"`
	EntryName            string   `json:"entryName"`
	MeanNormalizedPayout *float64 `json:"meanNormalizedPayout,omitempty"`
	PTop1                *float64 `json:"pTop1,omitempty"`
	PInMoney             *float64 `json:"pInMoney,omitempty"`
	Rank                 int      `json:"rank"`
}

// LabEvaluationEntryBid represents a single team bid within an evaluation entry.
type LabEvaluationEntryBid struct {
	TeamID     string  `json:"teamId"`
	SchoolName string  `json:"schoolName"`
	Seed       int     `json:"seed"`
	Region     string  `json:"region"`
	BidPoints  int     `json:"bidPoints"`
	Ownership  float64 `json:"ownership"`
}

// LabEvaluationEntryProfile represents a full entry profile with result and bids.
type LabEvaluationEntryProfile struct {
	EntryName            string                  `json:"entryName"`
	MeanNormalizedPayout *float64                `json:"meanNormalizedPayout,omitempty"`
	PTop1                *float64                `json:"pTop1,omitempty"`
	PInMoney             *float64                `json:"pInMoney,omitempty"`
	Rank                 int                     `json:"rank"`
	TotalBidPoints       int                     `json:"totalBidPoints"`
	Bids                 []LabEvaluationEntryBid  `json:"bids"`
}

// LabLeaderboardEntry represents a row from lab.model_leaderboard view.
type LabLeaderboardEntry struct {
	InvestmentModelID         string     `json:"investmentModelId"`
	ModelName                 string     `json:"modelName"`
	ModelKind                 string     `json:"modelKind"`
	NEntries                  int        `json:"nEntries"`
	NEntriesWithPredictions   int        `json:"nEntriesWithPredictions"`
	NEvaluations              int        `json:"nEvaluations"`
	NCalcuttasWithEntries     int        `json:"nCalcuttasWithEntries"`
	NCalcuttasWithEvaluations int        `json:"nCalcuttasWithEvaluations"`
	AvgMeanPayout             *float64   `json:"avgMeanPayout,omitempty"`
	AvgMedianPayout           *float64   `json:"avgMedianPayout,omitempty"`
	AvgPTop1                  *float64   `json:"avgPTop1,omitempty"`
	AvgPInMoney               *float64   `json:"avgPInMoney,omitempty"`
	FirstEvalAt               *time.Time `json:"firstEvalAt,omitempty"`
	LastEvalAt                *time.Time `json:"lastEvalAt,omitempty"`
}

// LabListModelsFilter for filtering investment models list.
type LabListModelsFilter struct {
	Kind *string
}

// LabListEntriesFilter for filtering entries list.
type LabListEntriesFilter struct {
	InvestmentModelID *string
	CalcuttaID        *string
	StartingStateKey  *string
}

// LabListEvaluationsFilter for filtering evaluations list.
type LabListEvaluationsFilter struct {
	EntryID           *string
	InvestmentModelID *string
	CalcuttaID        *string
}

// LabPagination for list operations.
type LabPagination struct {
	Limit  int
	Offset int
}

File: frontend/src/test/msw/handlers/lab.ts
import { http, HttpResponse } from 'msw';

const BASE = 'http://localhost:8080/api/v1';

const validModel = {
  id: 'model-1',
  name: 'test-model',
  kind: 'regression',
  paramsJson: { lr: 0.01 },
  notes: null,
  createdAt: '2026-01-01T00:00:00Z',
  updatedAt: '2026-01-01T00:00:00Z',
  nEntries: 5,
  nEvaluations: 3,
};

const validEntry = {
  id: 'entry-1',
  investmentModelId: 'model-1',
  calcuttaId: 'calc-1',
  gameOutcomeKind: 'kenpom',
  gameOutcomeParamsJson: {},
  optimizerKind: 'milp',
  optimizerParamsJson: {},
  startingStateKey: 'pre_tournament',
  hasPredictions: true,
  predictions: [
    {
      teamId: 'team-1',
      schoolName: 'Duke',
      seed: 1,
      region: 'East',
      predictedBidPoints: 15,
      expectedPoints: 50,
      expectedRoi: 2.5,
      rationalPoints: 10,
      edgePercent: 0.3,
    },
  ],
  bids: [
    {
      teamId: 'team-1',
      schoolName: 'Duke',
      seed: 1,
      region: 'East',
      bidPoints: 20,
      rationalPoints: 10,
      edgePercent: 0.5,
    },
  ],
  createdAt: '2026-01-01T00:00:00Z',
  updatedAt: '2026-01-01T00:00:00Z',
  modelName: 'test-model',
  modelKind: 'regression',
  calcuttaName: 'Test Pool',
  nEvaluations: 3,
};

const validEvaluation = {
  id: 'eval-1',
  entryId: 'entry-1',
  nSims: 1000,
  seed: 42,
  meanNormalizedPayout: 1.25,
  medianNormalizedPayout: 1.1,
  pTop1: 0.15,
  pInMoney: 0.6,
  ourRank: 2,
  simulatedCalcuttaId: null,
  createdAt: '2026-01-01T00:00:00Z',
  updatedAt: '2026-01-01T00:00:00Z',
  modelName: 'test-model',
  modelKind: 'regression',
  calcuttaId: 'calc-1',
  calcuttaName: 'Test Pool',
  startingStateKey: 'pre_tournament',
};

export const labHandlers = [
  http.get(`${BASE}/lab/models/leaderboard`, () => {
    return HttpResponse.json({
      items: [
        {
          investmentModelId: 'model-1',
          modelName: 'test-model',
          modelKind: 'regression',
          nEntries: 5,
          nEntriesWithPredictions: 5,
          nEvaluations: 3,
          nCalcuttasWithEntries: 2,
          nCalcuttasWithEvaluations: 2,
          avgMeanPayout: 1.25,
          avgMedianPayout: 1.1,
          avgPTop1: 0.15,
          avgPInMoney: 0.6,
          firstEvalAt: '2026-01-01T00:00:00Z',
          lastEvalAt: '2026-01-15T00:00:00Z',
        },
      ],
    });
  }),

  http.get(`${BASE}/lab/models/:name/calcutta/:calcuttaId/entry`, () => {
    return HttpResponse.json(validEntry);
  }),

  http.get(`${BASE}/lab/models/:id/pipeline/progress`, () => {
    return HttpResponse.json({
      modelId: 'model-1',
      modelName: 'test-model',
      activePipelineRunId: null,
      totalCalcuttas: 2,
      predictionsCount: 2,
      entriesCount: 2,
      evaluationsCount: 2,
      avgMeanPayout: 1.25,
      calcuttas: [
        {
          calcuttaId: 'calc-1',
          calcuttaName: 'Test Pool',
          calcuttaYear: 2025,
          stage: 'complete',
          status: 'succeeded',
          progress: 1.0,
          hasPredictions: true,
          hasEntry: true,
          hasEvaluation: true,
          entryId: 'entry-1',
          evaluationId: 'eval-1',
          meanPayout: 1.25,
          ourRank: 2,
        },
      ],
    });
  }),

  http.get(`${BASE}/lab/models/:id`, () => {
    return HttpResponse.json(validModel);
  }),

  http.get(`${BASE}/lab/evaluations/:id/entries`, () => {
    return HttpResponse.json({
      items: [
        {
          id: 'result-1',
          entryName: 'Test Entry',
          meanNormalizedPayout: 1.25,
          pTop1: 0.15,
          pInMoney: 0.6,
          rank: 1,
        },
      ],
    });
  }),

  http.get(`${BASE}/lab/evaluations/:id`, () => {
    return HttpResponse.json(validEvaluation);
  }),

  http.get(`${BASE}/lab/evaluations`, () => {
    return HttpResponse.json({ items: [validEvaluation] });
  }),

  http.get(`${BASE}/lab/entry-results/:id`, () => {
    return HttpResponse.json({
      entryName: 'Test Entry',
      meanNormalizedPayout: 1.25,
      pTop1: 0.15,
      pInMoney: 0.6,
      rank: 1,
      totalBidPoints: 100,
      bids: [
        {
          teamId: 'team-1',
          schoolName: 'Duke',
          seed: 1,
          region: 'East',
          bidPoints: 20,
          ownership: 0.5,
        },
      ],
    });
  }),

  http.post(`${BASE}/lab/models/:id/pipeline/start`, () => {
    return HttpResponse.json({
      pipelineRunId: 'run-1',
      nCalcuttas: 2,
      status: 'running',
    });
  }),

  http.post(`${BASE}/lab/pipeline-runs/:id/cancel`, () => {
    return HttpResponse.json({ status: 'cancelled' });
  }),
];
